### PR TITLE
Issue #291 better one line scope indentation

### PR DESCRIPTION
**Pull request by camsteffen**

### CODE REVIEW

**Comment by qstrahl** [comment]
@camsteffen Can you elaborate on what this change addresses? Also you'll need to rebase if you want it merged in

**Comment by davidchambers** [comment]
Presumably this solves the issue raised in #291:

>     if(1) return
>         console.log('foo')

**Comment by qstrahl** [comment]
@camsteffen It looks like you merged our `develop` branch into yours; you will have to undo that change with a `git reset --hard b246a506bd1bdfa0827a9861707ffb0b6934b3c5` and instead do a rebase with `git rebase refs/remotes/pangloss/develop` You will have to resolve the resultant merge conflict.

**Comment by camsteffen** [comment]
Thanks for helping with the rebase. Hopefully it is set to go now.

To recap the change, the regex now checks to see if there is a statement after the parenthesis for a `if|for|while` so that `if(foo) bar()` does not begin a one line scope. If a line ends with `else`, then it does begin a one line scope. This pattern could potentially break for something like `while("(foo)" === bar)` but that is an edge case.

**Comment by qstrahl** [comment]
I'm willing to try it out for now. We can revert if breakages come up.

**Comment by bounceme** [comment]
hey, some things are still being indented
```
function n (err) {
  if (err) return log(err)
    var act = 0
}
```

**Comment by camsteffen** [comment]
Unfortunately I think we have the best solution possible using regular expressions. It becomes an ultimatum between (1) falsely indent after `if(foo) bar()` or (2) falsely not indent after `if(foo())`. I prefer option 1, especially since it can be fixed with a semicolon: `if(foo) bar();`

File: indent/javascript.vim [hunk 1]
```diff
@@ -66,7 +66,7 @@ let s:continuation_regex = '\%([\\*+/.:]\|\%(<%\)\@<![=-]\|\W[|&?]\|||\|&&\|[^=]
 " TODO: this needs to deal with if ...: and so on
 let s:msl_regex = s:continuation_regex.'|'.s:expr_case
 
-let s:one_line_scope_regex = '\%(\<\%(if\|else\|for\|while\)\>\|=>\)[^{;]*' . s:line_term
+let s:one_line_scope_regex = '\%(\<else\>\|\<\%(if\|for\|while\)\>\s*(.*)\)' . s:line_term
```
Comment (bounceme) [inline indent/javascript.vim]: unexpectedly this seems to be working:
```
let s:one_line_scope_regex = '^[})\]]*\s*\(if\|else\|while\|try\|catch\|finally\|for\|else\s\+if\)\s*_\=$'
```
gotten from https://github.com/jiangmiao/simple-javascript-indenter/blob/master/indent/javascript.vim#L350

**Comment by bounceme** [comment]
never mind, there isn't indentation when the curly braces are omitted

**Comment by camsteffen** [comment]
can you give an example?

**Comment by bounceme** [comment]
```
  if (true)
  console.log("value");
```

**Comment by camsteffen** [comment]
Your example indents correctly for me...

**Comment by bounceme** [comment]
did you replace line 69 with
```
let s:one_line_scope_regex = '^[})\]]*\s*\(if\|else\|while\|try\|catch\|finally\|for\|else\s\+if\)\s*_\=$'
```

**Comment by camsteffen** [comment]
No, but now I'm confused. Why are we looking at this other line of code if it doesn't work?

**Comment by bounceme** [comment]
the indent script that it is copied from, jiangiao's, does correctly indent these different variations of ifs.maybe we can get a similar behavior by looking at other parts of that script to see what he did

**Comment by camsteffen** [comment]
Sorry, it doesn't solve the problem that I described before:
> It becomes an ultimatum between (1) falsely indent after `if(foo) bar()` or (2) falsely not indent after `if(foo())`. I prefer option 1, especially since it can be fixed with a semicolon: `if(foo) bar();`

Here's another way to look at it. Using regular expressions, there is no way to detect whether parenthesis are matching. It can't see if `if (` matches with `)` at the end of the line.

**Comment by bounceme** [comment]
well, it may be possible if we try with ```:h searchpair()```

**Comment by camsteffen** [comment]
Good find, I didn't know about that. Maybe I'll look into it when I have some time. It would be more than a one-liner fix.

**Comment by bounceme** [comment]
This is what I think would be easiest,though not elegant:
```
let s:one_line_scope_regex = '\%(\%(\<else\>\|\<\%(if\|for\|while\)\>\s*(\%([^()]\|[^()]*([^()]*)[^()]*\))\)\|=>\)' . s:line_term
```
This can be extended to allow nesting more than two pairs

**Comment by bounceme** [comment]
last comment showed two pairs nested,here's what 4 looks like...
```
let s:one_line_scope_regex = '\%(\%(\<else\>\|\<\%(if\|for\|while\)\>\s*(\%([^()]*\|[^()]*(\%([^()]*\|[^()]*(\%([^()]*\|[^()]*([^()]*)[^()]*\))[^()]*\))[^()]*\))\)\|=>\)' . s:line_term
```

**Pull request merged**